### DISCUSSION TITLE
ggplot2 plotting mindshare theories

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#41] [General]
A spot to think about when plotting tasks consumes relatively more or little mindshare.

Assertion: less mindshare is consumed when it feels like you can move one specification to another in one step for things that feel like they should happen in one step.  'feel like they should happen in one step' is pretty subjective of course.  

I think the following example is a fence case.  It kind of feels like you should be able to go from vanilla to ordered and it be top is largest in one step.  I think fastest, clearest way is fct_infreq and then fct_reverse...   

![image](https://github.com/user-attachments/assets/2267700b-603e-4904-9e68-8c80c1db4790)

cut off bit is 'fct_reverse'...  And yeah, I guess fct functions puts this discussion in larger tidyverse/base r manipulation context.

**Comment by yjunechoe** [comment]
I would love for this to be handled inside the scales somehow!

This will probably break too many things, but it'd be kinda fun if scales could retain some information about the original data. To the point about the bar ordering example, specifically, it'd be neat if we could express that as:

```r
penguins %>% 
  ggplot() +
  aes(y = species) +
  geom_bar() +
  scale_y_discrete(
    limits = ~ rev(levels(fct_infreq(.x)))
  )
```

Currently only the value of the limits itself is available, so you'd need to go back to the data to get the original frequency information:

```r
library(ggplot2)
library(forcats)
library(palmerpenguins)

penguins %>% 
  ggplot() +
  aes(y = species) +
  geom_bar() +
  scale_y_discrete(
    limits = rev(levels(fct_infreq(penguins$species)))
  )
```

![image](https://github.com/user-attachments/assets/b0d32903-0aea-48a0-b57a-8a64ed66527e)

Another usecase I'd come across before is what if you want a continuous axis to break at intervals specific to the data, like the quartiles? You can of course do:

```r
penguins %>% 
  ggplot(aes(bill_length_mm)) +
  geom_density() +
  scale_x_continuous(
    breaks = unname(quantile(penguins$bill_length_mm, na.rm = TRUE)),
    sec.axis = sec_axis(
      transform = identity,
      breaks = quantile(penguins$bill_length_mm, na.rm = TRUE)
    )
  )
#> Warning: Removed 2 rows containing non-finite outside the scale range (`stat_density()`).
```

![image](https://github.com/user-attachments/assets/e9944898-e9ac-429c-8a5a-fb8e947f3ed0)

But it'd be neat if you didn't have to reach back out to the original data to get `penguins$bill_length_mm`

**Reply by EvaMaeRey** [reply]
Yeah. from user prespecive + scale_y_infreq would probably feel right.. or geom_bar_infreq() 

Maybe ggplot_add could be used to reach back in and relate x and y, and write something that looks like scale_y_*?  y and x not being orthogonal makes things tricky!

**Reply by yjunechoe** [reply]
I think `{tidytext}` comes clost in venturing into that direction of "extension", with `reorder_within()` and `scale_x|y_reordered()`: https://juliasilge.github.io/tidytext/reference/reorder_within.html

**Reply by EvaMaeRey** [reply]
I've seen reorder_within() in ggcharts.  But wasn't aware of scale_y_reordered!

**Reply by teunbrand** [reply]
I get that this feels like it shouldn't be this difficult. And in the majority of plots, where you have a single layer, you'd be right. However, the problems start compounding when you have multiple layers that can have different data.  How does the scale know what to count? Both layers? The first layer? Only the global data?

**Reply by EvaMaeRey** [reply]
fct_infreq_rev()!

**Reply by EvaMaeRey** [reply]
I'm just seeing a y ordered in-freq within facet ordering question articulated here:  facethttps://github.com/tidyverse/forcats/issues/377  Very similar to the ggcharts discussion motivating the package (which solves the problem by wrapping): https://github.com/thomas-neitmann/ggcharts

**Reply by EvaMaeRey** [reply]
I think this could be a viable in-script, within-tidyverse method.  And could be slightly more concise if sep is dropped in `interaction` and `___` is replaced with '\\\\.`  

``` r
gapminder::gapminder %>% 
  filter(continent == "Americas") %>% 
  filter(year %in% c(1957, 2002)) %>% 
     ggplot() + 
     facet_wrap(~ year) + 
     aes(
         x = pop, 
         y = interaction(year, country, sep = "___") |> fct_reorder(pop)
        ) + 
     geom_col() 
```
![image](https://github.com/user-attachments/assets/3c960660-73cf-4482-830c-616b839b46ea)

``` r
last_plot() + 
  facet_wrap(~ year, scales = "free_y")
```

![image](https://github.com/user-attachments/assets/949d785f-d605-4f69-b95e-d12e42e4efc3)

``` r
last_plot() +
  scale_y_discrete(labels = function(x) str_remove(x, ".+___")) + # don't show year part of factor
  labs(y = NULL)
```

![image](https://github.com/user-attachments/assets/9afb98ba-f7c6-49e1-a2b0-045a0756f7d3)

**Reply by EvaMaeRey** [reply]
... or probably better, drop need for facet var in defining y: interaction(rank(pop), country, lex.order = T). 🙂 Just finding my way around interaction() args.

**Reply by EvaMaeRey** [reply]
> I get that this feels like it shouldn't be this difficult. And in the majority of plots, where you have a single layer, you'd be right. However, the problems start compounding when you have multiple layers that can have different data. How does the scale know what to count?

I'm on board w/ the logic above, but I am curious if it must be the case that computed variables cannot (?) be new discrete vars.  So I have a little trouble with '#> ! Discrete values supplied to continuous scale.'  If 'y' isn't specified, does it necessarily have to be the case that a discrete var can't be computed in the Stat?  For the rank-ordered bar chart, it would be sooo nice.  🥹

``` r
library(ggplot2)
library(dplyr)

compute_panel_rank <- function(data, scales){
  
  data|>
    mutate(y = interaction(ranked, rank(-x)))
  
}

gapminder::gapminder |>
  filter(year == 2002) |> 
  filter(continent == "Americas") |>
  ggplot() + 
       aes(x = pop, 
               ranked = country) +   # <<  the ranked variable, to be interacted with x to create an ordered factor to compute y
  geom_point(stat = ggproto("", Stat, compute_panel = compute_panel_rank))
#> Error:
#> ! Discrete values supplied to continuous scale.
#> ℹ Example values: Argentina.5, Bolivia.14, Brazil.2, Canada.6, and Chile.9
```

<sup>Created on 2025-03-03 with [reprex v2.1.1.9000](https://reprex.tidyverse.org)</sup>

**Reply by teunbrand** [reply]
I think you cannot have mapped computed discrete variables because it would mess up scale training, especially for position variables. Discrete position variables are continu-ized (like the opposite of discretise? is this a word?) before they are fed into stats, so upon arrival discrete scales should already know what levels they are representing.

**Reply by EvaMaeRey** [reply]
I think you've answered this a few times before. 🙏  And maybe this example is like one where the answer to the question ... 'If 'y' isn't specified, does it necessarily have to be the case that a discrete var can't be computed in the Stat?' ... that really points to computed vars logically if factors were allowed.  In this case, the in-facet computation would give you a certain label-numeric factor set, that wouldn't combine *across* facets (within facet Portugal.2.2005 |> as.numeric() is 2, and Brazil.2.2006 |> as.numeric() is also 2); layer_data is *across* facets, so kind of a mismatch?

**Comment by EvaMaeRey** [comment]
using the options and stat_region() as discussed in #48, to look look at some brain regions (over here because of the mindshare topic, though it somewhat became more specific about bar chart ordering)
![image](https://github.com/user-attachments/assets/177273a4-7602-46fa-8d7f-3cf518b9044b)

**Reply by EvaMaeRey** [reply]
for the stamp_* functions (think annotation), the required aes shouldn't actually take effect.  I haven't successfully addressed that so using region = 1 to satisfy the requirement.

**Comment by EvaMaeRey** [comment]
Cross reference to forcats discussion, request for fct_reorder_within... https://github.com/tidyverse/forcats/issues/377#issue  And yet another idea as to how make the build feel cleaner (?), `aes_y_by_x` - a little unorthodox.



``` r
library(tidyverse)

aes_y_by_x <- function(x, y){
  
  aes(x = {{x}}, 
      y = interaction(rank({{x}}), {{y}}, lex.order = T))
  
}

gapminder::gapminder |>
  filter(year %in% c(1967, 2007)) |> 
  filter(continent == "Americas") |>
  ggplot() + 
  aes_y_by_x(x = pop, y = country) +
  geom_col() + 
  facet_wrap(facets = vars(year), 
             scales = "free_y") 
```

![](https://i.imgur.com/2pnVUMD.png)

``` r

last_plot() +
  scale_y_discrete(labels = function(x) str_remove(x, "^\\d+?\\."))
```

![](https://i.imgur.com/LaxnYbK.png)

``` r

last_plot() +
  ggplyr::data_slice_max(pop, by = year)
```

![](https://i.imgur.com/XjS6vXc.png)

<sup>Created on 2025-08-23 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Reply by EvaMaeRey** [reply]
https://github.com/tidyverse/forcats/issues/377#issuecomment-3218255782  @abiyug, if you'd like to move the conversation here.  I guess ggplot2 has a lot of it's own grouping logic, so I think for this reason, probably the group information doesn't (shouldn't?) persist from tibble -> plot environment. 

 I was glad that the ggcharts::bar_chart() documentation exposes all of the how-to in the documentation (if as motivation *not* to use a more ggplot2-syntaxed method).

**Comment by abiyug** [comment]
Hey everyone! We’ve been exploring a common challenge with `ggplot2`: the need to reorder factor levels within facets (e.g., by a value like **value** in `facet_wrap`) isn’t natively supported, requiring pre-processing with dplyr and forcats. This adds cognitive load, especially since ordering within multi-plot setups feels like a natural step for many users. _I'm considering a feature request_ for an `order_by` argument in` facet_wrap` (e.g., `facet_wrap(~ category, order_by = value, direction = "desc")`) to simplify this, or alternatively, a _**ggplot2 extension**_ to handle it.
 I’d love your input, whether you’ve faced this issue, thoughts on the proposal, or ideas for prototyping an extension—to ensure this resonates with the community and moves forward effectively.

**Reply by EvaMaeRey** [reply]
> `facet_wrap(~ category, order_by = value, direction = "desc")`  

I think a potential issue with this might be that it would be ambiguous which aesthetic `order_by` is directed at (y, fill, x?) ... and more generally that facets are kind of orthogonal to the aesthetic mappings (but there is the free x and free y specifications ...).  Maybe more fundamentally, there's the early scale training issue @teunbrand 

> Discrete position variables are continuized (like the opposite of discretize? is this a word?) before they are fed into stats, so upon arrival discrete scales should already know what levels they are representing.
above https://github.com/ggplot2-extenders/ggplot-extension-club/discussions/41#discussioncomment-12378233

But I do feel like...

``` r
ggplot(data) + 
     aes(y) + 
     geom_bar_ordered() # ggplot_add violence here where 1) y factor is reordered by x (count or other) and interacted with the rank, 2) scale is also included to hide the rank information, and also reversed so highest is on top.

last_plot() + 
facet_wrap()
```

... would be easy to use

**Reply by teunbrand** [reply]
I've sometimes mused about how feasible it would be to implement facets as a discrete scale. Essentially the "scale" transformation takes the variables you declare and outputs a `PANEL` variable. You'd be able to set `limits` (i.e. universe of possible values), `breaks` (the values that are displayed & in what order) and `labels`, which would supplant the esoteric `facet_*(labeller)` mechanism. Especially the `limits`/`breaks` functionality is currently lacking. The `palette` is then mapping the `PANEL` variable to a 2D location in a grid, and the `guide` would be the strips.

**Reply by EvaMaeRey** [reply]
like aes(facets = year) for wrapping? aes(rows = sex, cols = island) for grid??

**Reply by teunbrand** [reply]
That is kinda where the idea breaks down a lot of the time. There is no good mechanism to declare novelty aesthetics for existing layers.

**Reply by EvaMaeRey** [reply]
I think I don't know what a novelty aesthetic is (like not part of the cannon found in 'base' ggplot2, e.g. x, y, color, fill, linewidth)

**Reply by teunbrand** [reply]
My intended meaning was 'an aesthetic that is not (currently) part of a Geom or Stat's available aesthetics'.

**Reply by EvaMaeRey** [reply]
I put a sketch of `geom_col_ranked` in exploratory 'ggpro' https://github.com/EvaMaeRey/ggpro which is chart-type embracing but tries to do data + mapping + layer syntax.  Not super flexible, but I would use this.  @abiyug, I'd be curious what you think about the api.  It might be that what @teunbrand is thinking might be more principled and general.

``` r
ggplot(americas_pop_2years) + 
  aes(x = pop_millions, y = country) +
  geom_col_ranked() +
  facet_wrap(~year, scales = "free_y")
```

<img width="672" height="480" alt="image" src="https://github.com/user-attachments/assets/10cb6af3-65f7-494d-bfc6-754126c31023" />

where `geom_col_ranked` is defined with a remapping, geom_col, scale and some other stuff that felt right (labs(y = NULL, theme moves).

``` r
geom_col_ranked <- function(...){
  
  list(geom_col(...),
       re_map_aes_y_by_x(),  
       scale_y_remove_rank_rev(),    
       theme(panel.grid.minor.y = element_blank(),
                   panel.grid.major.y = element_blank()),
       labs(y = NULL))
  
}
```

**Comment by abiyug** [comment]
Just to restate our objective: "when rendering **_multiple plots_**, `ggplot2` should auto-magically order bars per facet". @EvaMaeRey Your `geom_col_ranked ` achieves the ordered plot when using `facet_wrap`, placing the ordering per panel on `geom_col` as a workaround within ggplot2’s current framework. This approach is similar to the short-term fix I proposed in our conversation trail from issue #377, offering a tailored solution for specific scenarios but lacking the flexibility to adapt to diverse use cases. In contrast, I’d like to share a function inspired by @teunbrand’s input, `facet_scale`, which introduces a dynamic and reusable solution by treating facets as a discrete scale with a **PANEL** variable. This could potentially motivate the adoption of a new feature into ggplot2’s core, enhancing its capability to handle ordering natively. Let me know your thoughts on this approach ...

```
# Objective: To develop a function that reorders the y-variable within each facet group based on a specified order_by variable, with the intent of prototyping a scale-based faceting system inspired by Teunbrand's comments.  This aims to enable dynamic ordering (via limits, breaks, and labels) and prepare the groundwork for a potential ggplot2 core feature that auto-magically handles bar ordering per facet, reducing the need for pre-wrangling.

library(dplyr)
library(forcats)
library(ggplot2)

# Preliminary facet_scale function
facet_scale <- function(data, facets, y_var, order_by, direction = "desc", limits = NULL, breaks = NULL) {
  facet_var <- dplyr::ensym(facets)
  y_var <- dplyr::ensym(y_var)
  order_by <- dplyr::ensym(order_by)
  
  data %>%
    dplyr::group_by(!!facet_var) %>%
    dplyr::mutate(
      rank_value = if (direction == "desc") -rank(!!order_by, ties.method = "first") 
                   else rank(!!order_by, ties.method = "first"),
      !!y_var := forcats::fct_reorder(as.character(!!y_var), rank_value)
    ) %>%
    dplyr::ungroup() %>%
    dplyr::select(-rank_value) %>%
    dplyr::mutate(PANEL = forcats::fct_relevel(!!facet_var, breaks %||% unique(!!facet_var)))
}

# Example data
df <- tibble::tribble(
        ~country, ~category, ~value,
         "Sudan", "biofuel",   0.94,
  "South Africa", "biofuel",    0.4,
        "Zambia", "biofuel",   0.88,
  "South Africa",    "coal", 192.73,
         "Egypt",    "coal", 179.22,
       "Morocco",    "coal",  33.12,
)


# Apply the function
df_scaled <- facet_scale(df, facets = category, y_var = country, order_by = value, direction = "desc")

# ggplot script using df_scaled
ggplot(df_scaled, aes(x = value, y = country)) +
  geom_col() +
  facet_wrap(~ category, scales = "free") +  # Using category for now, as PANEL is preliminary
  scale_y_discrete(limits = rev)  # Reverse y-axis to show highest value on top
```

**Reply by EvaMaeRey** [reply]
Hi @abiyug, I like the desire to keep us focused and on task! 😅  Our vibe has been more 'muse and meander'.   I believe @teunbrand  describes a pretty major overhaul to ggplot.  Is that right (?).  'strips are the guides.  being able to define the limits and breaks...' sounds super!  `facet_scale()` looks worth pursing meanwhile!

**Reply by abiyug** [reply]
Thanks for the feedback! I’m glad  you think `facet_scale` is a good alternative. My initial aim was for `ggplot2` to natively order bars per facet for `geom_col` in multiplots, like _Tableau_ does without extra steps. I may have underestimated the core changes needed, but Gina’s `geom_col_ranked` and my `facet_scale` helper prove pre-wrangling can order bars. Perhaps the `ggplot2` maintainers could add this to their roadmap, given the intuitive experience it offers.

**Reply by EvaMaeRey** [reply]
One thing that falls short with the geom_col_ranked is that it doesn't do the counting for you.  geom_bar_ranked would be nice - but I'm not sure it's possible using the same approach - but haven't given it much thought.
